vitrina/orgs/services.py

```diff
@@ -1,16 +1,150 @@
+            # user can manage dataset/request
+            if isinstance(obj, (Dataset, Request)):
+                ct = ContentType.objects.get_for_model(obj)
+                if representative_exists(user, ct, obj.pk):
+                    return True
+                elif representative_exists(user, organization_ct, obj.organization.pk):
+                    return True
+
+            # if user is organization coordinator, user can manage organization
+            if isinstance(obj, Organization) and \
+                    representative_exists(user, organization_ct, obj.pk, Representative.COORDINATOR):
+                return True
+
+            # if user is organization representative, user can manage organization representative
+            if isinstance(obj, Representative) and isinstance(obj.content_object, Organization) and \
+                    representative_exists(user, organization_ct, obj.object_id, Representative.COORDINATOR):
+                return True
+
+            # if user is dataset/request and dataset/request organization coordinator,
+            # user can manage dataset/request representative
+            if isinstance(obj, Representative) and isinstance(obj.content_object, (Dataset, Request)):
+                ct = ContentType.objects.get_for_model(obj.content_object)
+                if representative_exists(user, ct, obj.object_id, Representative.COORDINATOR) and \
+                    representative_exists(user, organization_ct,
+                                          obj.content_object.organization.pk, Representative.COORDINATOR):
+                    return True
+                # if user is only dataset/request coordinator, user can manage dataset/request representative
+                # if it is from organizations where user is coordinator
+                elif representative_exists(user, ct, obj.object_id, Representative.COORDINATOR):
+                    # TODO: need to check if representative is from specific organization
+                    # if we don't remove organization attribute from representative, then need to check if
+                    # obj.organization.pk in Representative.objects.filter(
+                    #                         user=user,
+                    #                         object_id=obj.object_id,
+                    #                         content_type=organization_ct,
+                    #                         role=Representative.COORDINATOR
+                    #                 ).values_list('organization_id', flat=True)
+                    pass
+
+            # if user is dataset or dataset organization manager or coordinator, user can manage dataset distribution
+            if isinstance(obj, DatasetDistribution):
+                if representative_exists(user, dataset_ct, obj.dataset.pk):
+                    return True
+                if representative_exists(user, organization_ct, obj.dataset.organization.pk):
+                    return True
```

Comment:
Too much code and too many places for possible errors. Maybe it is possible to refactor this code into a simpler, declarative version?

Declaration of access rights could look like this:

```python
acl = {
    (Organization, Action.UPDATE): [Role.COORDINATOR],
    (Representative, Action.CREATE): [Role.COORDINATOR],
    (Representative, Action.UPDATE): [Role.COORDINATOR],
    (Representative, Action.DELETE): [Role.COORDINATOR],
    (Dataset, Action.CREATE): [Role.COORDINATOR, Role.MANAGER],
    (Dataset, Action.UPDATE): [Role.COORDINATOR, Role.MANAGER],
    (Dataset, Action.DELETE): [Role.COORDINATOR, Role.MANAGER],
    (DatasetDistribution, Action.CREATE): [Role.COORDINATOR, Role.MANAGER],
    (DatasetDistribution, Action.UPDATE): [Role.COORDINATOR, Role.MANAGER],
    (DatasetDistribution, Action.DELETE): [Role.COORDINATOR, Role.MANAGER],
    (Request, Action.CREATE): [ALL],
    (Request, Action.UPDATE): [AUTHOR],
    (Request, Action.DELETE): [AUTHOR],
    (Project, Action.CREATE): [ALL],
    (Project, Action.UPDATE): [AUTHOR],
    (Project, Action.DELETE): [AUTHOR],
}
```

Having this declarative table of access rights, it is easy to add new rules and there is less room for error.

Also, there are two special roles:

- `ALL` - any authenticated user.
- `AUTHOR` - user who is author of a model instance.

Now, it is easy to implement a generalized access checks, something like this (this is a pseudo code, might not work):

```python
if not user.is_authenticated:
    return False

if isinstance(obj, Type):
    model = obj
    nodes = get_parents(parent)
else:
    model = type(obj)
    nodes = get_parents(obj)

where = []
for role in acl[(model, action)]:
    if role == ALL:
        return True
    else:
        for node in nodes:
            if role == AUTHOR:
                if is_author(user, node):
                    return True
            else:
                where.append(Q(content_object=node, role=role))

where = functools.reduce(operator.or_, where)

return Representative.objects.filter(where, user=user).exists()
```

This requires tow additional functions:

- `is_author(user, obj)` - returns true if `user` is author of given `obj`.
- `get_parents(obj)` - returns list of parents of `obj`, including `obj` itself. I guess in order to implement this, you can add `get_acl_parents` method to each model, which in turn would call `get_acl_parents` of its parent.